### PR TITLE
gRPC e-book Chapter 4

### PR DESCRIPTION
**Pull request by ThatRendle**
Chapter 4 Comparing gRPC to WCF

### CODE REVIEW

**Comment by markrendle** [comment]
Ready for review. cc @glennc @saurabh500 @mconnew @JamesNK @JunTaoLuo @mairaw

File: docs/architecture/grpc-for-wcf-developers/error-handling.md [hunk 1]
```diff
@@ -7,5 +7,61 @@ ms.date: 09/02/2019
...
+Unfortunately, the current version of gRPC lacks the sophistication found with WCF and only has limited built-in error handling based on simple status codes and metadata. The following table is a quick guide to the most commonly-used status codes.
+
+| Status Code | Problem |
+| ----------- | ------- |
+| GRPC_STATUS_UNIMPLEMENTED | Method hasn’t been written |
+| GRPC_STATUS_UNAVAILABLE | Problem with the whole service |
+| GRPC_STATUS_UNKNOWN | Invalid response |
+| GRPC_STATUS_INTERNAL | Problem with encoding/decoding |
+| GRPC_STATUS_UNAUTHENTICATED | Authentication Failed |
+| GRPC_STATUS_PERMISSION_DENIED | Authorization Failed |
```
Comment (JamesNK) [inline docs/architecture/grpc-for-wcf-developers/error-handling.md]: I've noticed that Cancelled is commonly used around HTTP call fails.

File: docs/architecture/grpc-for-wcf-developers/types-of-rpc.md [hunk 2]
```diff
@@ -7,5 +7,255 @@ ms.date: 09/02/2019
...
+Here is an example of a server streaming RPC.
+
+```protobuf
+service ThingStreamer {
+    rpc Subscribe(SubscribeRequest) returns (stream Thing);
+}
+```
+
+```csharp
+{
```
Comment (JamesNK) [inline docs/architecture/grpc-for-wcf-developers/types-of-rpc.md]: nit: remove these top level braces

File: docs/architecture/grpc-for-wcf-developers/types-of-rpc.md [hunk 3]
```diff
@@ -7,5 +7,255 @@ ms.date: 09/02/2019
...
+    public class ThingStreamerService : ThingStreamer.ThingStreamerBase
+    {
+        public override async Task Subscribe(SubscribeRequest request,
+            IServerStreamWriter<Thing> responseStream,
+            ServerCallContext context)
+        {
+            while (!context.CancellationToken.IsCancellationRequested)
+            {
+                var thing = await _things.GetNext();
```
Comment (JamesNK) [inline docs/architecture/grpc-for-wcf-developers/types-of-rpc.md]: Where does _things come from? IMO simplify with declaring an int outside the loop and increase the returned string:

```cs
await responseStream.WriteAsync(new Thing { name = $"Thing {i}" });
```

File: docs/architecture/grpc-for-wcf-developers/types-of-rpc.md [hunk 4]
```diff
@@ -7,5 +7,255 @@ ms.date: 09/02/2019
...
+                await responseStream.WriteAsync(new Thing { name = thing.Name});
+            }
+        }
+    }
+}
+```
+
+```csharp
+public async Task GetThingsAsync(CancellationToken token)
```
Comment (JamesNK) [inline docs/architecture/grpc-for-wcf-developers/types-of-rpc.md]: What is this for?

File: docs/architecture/grpc-for-wcf-developers/types-of-rpc.md [hunk 5]
```diff
@@ -7,5 +7,255 @@ ms.date: 09/02/2019
...
+{
+    var request = new SubscribeRequest();
+    var response = _thingsClient.Subscribe(request);
+
+    while (await response.ResponseStream.MoveNext(token))
```
Comment (JamesNK) [inline docs/architecture/grpc-for-wcf-developers/types-of-rpc.md]: IMO `ReadAllAsync()` looks nicer. Prefer it when reading from streams.

File: docs/architecture/grpc-for-wcf-developers/types-of-rpc.md [hunk 6]
```diff
@@ -7,5 +7,255 @@ ms.date: 09/02/2019
...
+        _logger = logger;
+    }
+
+    public override async Task<CompletedResponse> OpenConnection(IAsyncStreamReader<Thing> requestStream, ServerCallContext context)
+    {
+        while (await requestStream.MoveNext(context.CancellationToken))
+        {
+            _logger.LogInformation(requestStream.Current.Description);
+        }
+        return new ConnectionClosedResponse();
```
Comment (JamesNK) [inline docs/architecture/grpc-for-wcf-developers/types-of-rpc.md]: Use known Empty object?

File: docs/architecture/grpc-for-wcf-developers/types-of-rpc.md [hunk 7]
```diff
@@ -7,5 +7,255 @@ ms.date: 09/02/2019
...
+    {
+        while (await _stream.ResponseStream.MoveNext(token))
+        {
+            Console.WriteLine(_stream.ResponseStream.Current.Text);
+        }
+    }
+
+    public async ValueTask DisposeAsync()
+    {
+        _cancellationTokenSource.Cancel();
```
Comment (JamesNK) [inline docs/architecture/grpc-for-wcf-developers/types-of-rpc.md]: Is the cancellation token required? Completing the client request stream will cause the RequestStream.MoveNext on the server to return false, and the gRPC method to finish on the server. Then ResponseStream.MoveNext on the client will return false and the read task will complete.

It is a cleaner way to finish the gRPC call compared to cancellation.

```cs
        await _stream.RequestStream.CompleteAsync();
        await _readTask;
        _stream.Dispose();
```

File: docs/architecture/grpc-for-wcf-developers/wcf-bindings.md [hunk 8]
```diff
@@ -1,11 +1,37 @@
...
 ms.date: 09/02/2019
 ---
 
 # WCF bindings and transports
 
+WCF has lots of different built in *bindings* specifying different network protocols, wire formats and other implementation details. gRPC effectively has just one network protocol and one wire format. However, you are still likely to discover that gRPC offers the best solution in most cases. What follows is a short discussion about the most relevant WCF bindings and how they compare to their equivalent in gRPC.
+
+## NetTCP
+
+WCF's NetTCP binding allows for persistent connections, small messages and two-way messaging but only works between .NET clients and servers. gRPC allows the same functionality but based on an open standard. gRPC has almost all the features of WCF NetTCP binding, although not always in the same way. For example, in WCF, encryption is controlled through configuration and handled in the framework. In gRPC, encryption is achieved at the connection level using HTTP/2 over SSL/TLS.
```
Comment (JamesNK) [inline docs/architecture/grpc-for-wcf-developers/wcf-bindings.md]: > SSL/TLS

Hey @blowdart how do you feel about HTTP/2 connection encryption being called "SSL/TLS"? Would you prefer just "TLS"?

File: docs/architecture/grpc-for-wcf-developers/types-of-rpc.md (hunk 4 above)
Comment (markrendle) [reply docs/architecture/grpc-for-wcf-developers/types-of-rpc.md]: Added line explaining client code.

File: docs/architecture/grpc-for-wcf-developers/wcf-bindings.md (hunk 8 above)
Comment (blowdart) [reply docs/architecture/grpc-for-wcf-developers/wcf-bindings.md]: Just TLS. It's not SSL.

File: docs/architecture/grpc-for-wcf-developers/comparing-wcf-services-to-grpc.md [hunk 9]
```diff
@@ -1,11 +1,69 @@
 ---
-title: Comparing WCF Services to gRPC - gRPC for WCF Developers
+title: Comparing WCF to gRPC - gRPC for WCF Developers
 description: TO BE WRITTEN
```
Comment (mairaw) [inline docs/architecture/grpc-for-wcf-developers/comparing-wcf-services-to-grpc.md]: still missing

File: docs/architecture/grpc-for-wcf-developers/error-handling.md [hunk 10]
```diff
@@ -7,5 +7,62 @@ ms.date: 09/02/2019
```
Comment (mairaw) [inline docs/architecture/grpc-for-wcf-developers/error-handling.md]: description here missing too

File: docs/architecture/grpc-for-wcf-developers/wcf-bindings.md (hunk 8 above)
Comment (JamesNK) [reply docs/architecture/grpc-for-wcf-developers/wcf-bindings.md]: @markrendle Do a find/replace on "SSL/TLS" to "TLS" across all files. Worth mentioning once somewhere that TLS is the successor to SSL.

File: docs/architecture/grpc-for-wcf-developers/types-of-rpc.md [hunk 11]
```diff
@@ -7,5 +7,258 @@ ms.date: 09/02/2019
```
Comment (mairaw) [inline docs/architecture/grpc-for-wcf-developers/types-of-rpc.md]: I'd suggest we name this article rpc-types.md to avoid unnecessary small words (another one of our rules)

File: docs/architecture/grpc-for-wcf-developers/types-of-rpc.md [hunk 12]
```diff
@@ -7,5 +7,258 @@ ms.date: 09/02/2019
...
+```
+
+> [!NOTE]
+> Server streaming RPCs are useful for subscription-style services, and also for sending very large datasets when it would be inefficient or impossible to build the entire dataset in memory. However, streaming responses is not as fast as sending `repeated` fields in a single message, so as a rule streaming should not be used for small datasets.
+
+### Differences to WCF
+
+A WCF duplex service uses a client callback interface that can have multiple methods. A gRPC server streaming service can only send messages over a single stream. If you need multiple methods, use a message type with either [an Any field or a oneof field](protobuf-any-oneof.md) to send different messages, and write code in the client to handle them.
+
+In WCF, the one-way method called from the client returns immediately, but the `ServiceContract` class with the session is kept alive until the connection is terminated. In gRPC, the Task returned by the implementation method should not complete until the connection is closed.
```
Comment (mairaw) [inline docs/architecture/grpc-for-wcf-developers/types-of-rpc.md]: ServiceContract here refers to this class? https://docs.microsoft.com/en-us/dotnet/api/system.servicemodel.servicecontractattribute?view=netframework-4.8

File: docs/architecture/grpc-for-wcf-developers/types-of-rpc.md [hunk 13]
```diff
@@ -7,5 +7,258 @@ ms.date: 09/02/2019
...
+
+## WCF one-way operations and gRPC client streaming
+
+WCF provides One-Way operations (marked with `[OperationContract(IsOneWay = true)]`) that the client can call without waiting for a response. gRPC service methods always return a response, even if it is empty, and the client should always `await` that response. For "fire-and-forget" style messaging in gRPC, you can create a Client Streaming service.
```
Comment (mairaw) [inline docs/architecture/grpc-for-wcf-developers/types-of-rpc.md]: await probably shouldn't be in backticks here because of localization.

File: docs/architecture/grpc-for-wcf-developers/wcf-bindings.md [hunk 14]
```diff
@@ -1,11 +1,37 @@
 ---
```
Comment (mairaw) [inline docs/architecture/grpc-for-wcf-developers/wcf-bindings.md]: This file is also included in #14303. What's the correct chapter for this? We've already done some edits on that other PR.

File: docs/architecture/grpc-for-wcf-developers/wcf-endpoints-and-grpc-methods.md [hunk 15]
```diff
@@ -7,5 +7,36 @@ ms.date: 09/02/2019
```
Comment (mairaw) [inline docs/architecture/grpc-for-wcf-developers/wcf-endpoints-and-grpc-methods.md]: another file name that should be renamed. drop the and from it.

File: docs/architecture/grpc-for-wcf-developers/ws-protocols.md [hunk 16]
```diff
@@ -0,0 +1,32 @@
...
+ms.date: 09/02/2019
+---
+
+# WS-* protocols
+
+One of the real benefits of working with WCF was that it supported many of the existing *WS-\** standard protocols. In this section, we will briefly cover how these same WS-* protocols are managed in gRPC and, where there is no alternative, discuss what options are available.
+
+## Metadata Exchange - WS-Policy, WS-Discovery, etc
+
+SOAP services expose Web Services Description Language (WSDL) schema documents with information about data formats, operations, communication options etc. This could be used to generate Client code.
```
Comment (mairaw) [inline docs/architecture/grpc-for-wcf-developers/ws-protocols.md]: This could be used -> This schema could be used?

File: docs/architecture/grpc-for-wcf-developers/ws-protocols.md [hunk 17]
```diff
@@ -0,0 +1,32 @@
...
+
+gRPC works best when servers and clients are generated from the same `.proto` files, but a server reflection optional extension does provide a way to expose dynamic information from a running server. See the [Grpc.Reflection Nuget package](https://nuget.org/packages/Grpc.Reflection) and [the documentation](https://github.com/grpc/grpc/blob/master/doc/csharp/server_reflection.md)
+
+## Security – WS-Security, WS-Federation, XML Encryption, etc
+
+We will cover security, authentication and authorization in much more detail in Chapter 6. But it is worth noting here that, unlike WCF, gRPC does not support WS-Security, WS Federation, or XML Encryption. Even so, gRPC offers greater security than WCF because all gRPC communications are automatically encrypted because of HTTP/2 TLS, and can use X509 certificates for mutual client/server authentication.
```
Comment (mairaw) [inline docs/architecture/grpc-for-wcf-developers/ws-protocols.md]: link to chapter 6?

File: docs/architecture/grpc-for-wcf-developers/wcf-endpoints-and-grpc-methods.md [hunk 18]
```diff
@@ -7,5 +7,41 @@ ms.date: 09/02/2019
...
+## OperationContract properties
+
+The [OperationContract](xref:System.ServiceModel.OperationContractAttribute) attribute has properties to control or refine how it works. gRPC methods don’t offer this type of control. The following table sets out those `OperationContract` properties and how the functionality they specify is (or isn’t) dealt with in gRPC:
+
+| `OperationContract` property | gRPC                                             |
+| ---------------------------- | ------------------------------------------------ |
+| <xref:System.ServiceModel.OperationContractAttribute.Action>             | SOAP-related, no meaning in gRPC. |
+| <xref:System.ServiceModel.OperationContractAttribute.AsyncPattern>       | All gRPC service methods return `Task` objects. |
+| <xref:System.ServiceModel.OperationContractAttribute.HasProtectionLevel> | No message encryption; network encryption handled at the transport layer (TLS over HTTP/2). |
+| <xref:System.ServiceModel.OperationContractAttribute.IsInitiating>       | See gRPC streaming. |
```
Comment (mairaw) [inline docs/architecture/grpc-for-wcf-developers/wcf-endpoints-and-grpc-methods.md]: see where?

File: docs/architecture/grpc-for-wcf-developers/ws-protocols.md [hunk 19]
```diff
@@ -0,0 +1,32 @@
...
+
+## WS-ReliableMessaging
+
+gRPC transport provides similar guarantees to WCF Reliable Sessions without any need for additional configuration or code.
+
+## WS-Transaction, WS-Coordination
+
+WCF’s implementation of distributed transactions uses Windows’ Microsoft Distributed Transaction Coordinator or MSDTC. It works with resource managers that specifically support it, like SQL Server, MSMQ, Windows file systems, etc. In the modern, open, cross-platform microservices world there is no equivalent. See Appendix A for a discussion of transactions.
```
Comment (mairaw) [inline docs/architecture/grpc-for-wcf-developers/ws-protocols.md]: link to appendix A?

**Review by mairaw — COMMENTED** [review]
Made some edits. Have some additional questions and comments.

File: docs/architecture/grpc-for-wcf-developers/types-of-rpc.md (hunk 12 above)
Comment (markrendle) [reply docs/architecture/grpc-for-wcf-developers/types-of-rpc.md]: Yes

File: docs/architecture/grpc-for-wcf-developers/wcf-bindings.md (hunk 14 above)
Comment (markrendle) [reply docs/architecture/grpc-for-wcf-developers/wcf-bindings.md]: Yeah, it shouldn't be in this one.

File: docs/architecture/grpc-for-wcf-developers/metadata.md [hunk 20]
```diff
@@ -7,5 +7,51 @@ ms.date: 09/02/2019
 
 # Metadata
 
+It's possible to add generic key/value headers to WCF messages using an <xref:System.ServiceModel.OperationContextScope> and the <xref:System.ServiceModel.OperationContext.OutgoingMessageHeaders?displayProperty=nameWithType> property.
+
+gRPC calls and responses can also include metadata similar to HTTP headers. These are mostly invisible to gRPC itself and are passed through to be processed by your application code or middleware. Metadata is represented as key/value pairs where the key is a string and the value is either a string or binary data. You don’t need to specify metadata in the `.proto` file.
```
[the remark this answers is not in the document]
Comment (markrendle) [reply docs/architecture/grpc-for-wcf-developers/metadata.md]: Added an opening paragraph defining metadata.

File: docs/architecture/grpc-for-wcf-developers/types-of-rpc.md (hunk 12 above)
Comment (markrendle) [reply docs/architecture/grpc-for-wcf-developers/types-of-rpc.md]: Added link

File: docs/architecture/grpc-for-wcf-developers/wcf-bindings.md (hunk 14 above)
Comment (markrendle) [reply docs/architecture/grpc-for-wcf-developers/wcf-bindings.md]: Removed

File: docs/architecture/grpc-for-wcf-developers/wcf-endpoints-and-grpc-methods.md (hunk 18 above)
Comment (markrendle) [reply docs/architecture/grpc-for-wcf-developers/wcf-endpoints-and-grpc-methods.md]: Added link to relevant section

File: docs/architecture/grpc-for-wcf-developers/metadata.md [hunk 21]
```diff
@@ -7,5 +7,53 @@ ms.date: 09/02/2019
 
 # Metadata
 
+"Metadata" refers to additional data that may be useful while processing requests and responses but is not part of the actual application data. Metadata might include authentication tokens, request identifiers and tags for monitoring purposes, or information about the data like the number of records in a dataset.
```
Comment (mconnew) [inline docs/architecture/grpc-for-wcf-developers/metadata.md]: Consider comparing to WCF custom SOAP headers so that devs have an answer to "How do I map my usage of custom SOAP headers to gRPC?"

File: docs/architecture/grpc-for-wcf-developers/metadata.md [hunk 22]
```diff
@@ -7,5 +7,53 @@ ms.date: 09/02/2019
...
+
+It's possible to add generic key/value headers to WCF messages using an <xref:System.ServiceModel.OperationContextScope> and the <xref:System.ServiceModel.OperationContext.OutgoingMessageHeaders?displayProperty=nameWithType> property.
```
Comment (mconnew) [inline docs/architecture/grpc-for-wcf-developers/metadata.md]: There's also IncomingMessageHeaders. You can also get to headers via Message.Properties in places like message inspectors. Might be simpler to just reference the WCF MessageProperties class. If someone is using it, they should know what this is anyway.

File: docs/architecture/grpc-for-wcf-developers/rpc-types.md [hunk 23]
```diff
@@ -0,0 +1,268 @@
...
+    var thing = await _thingsClient.GetAsync(new GetThingRequest { ThingId = thingId });
+    Console.WriteLine($"{thing.Name}");
+}
+```
+
+As you can see, implementing a gRPC unary RPC service method is very similar to implementing a WCF operation, except that with gRPC you override a base class method instead of implementing an interface. Note that on the server, gRPC base methods always return a <xref:System.Threading.Tasks.Task%601>, although the client provides both async and blocking methods to call the service.
+
+## WCF duplex, one-way to client
+
+WCF applications (with certain bindings) can create a persistent connection between client and server, and the server can asynchronously send data to the client until the connection is closed, using a *callback interface*.
```
Comment (mconnew) [inline docs/architecture/grpc-for-wcf-developers/rpc-types.md]: Maybe be a bit more concrete and say that they do this by setting the callback service contract interface type on ServiceContract.CallbackContract.

File: docs/architecture/grpc-for-wcf-developers/rpc-types.md [hunk 24]
```diff
@@ -0,0 +1,268 @@
...
+```
+
+> [!NOTE]
+> Server streaming RPCs are useful for subscription-style services, and also for sending very large datasets when it would be inefficient or impossible to build the entire dataset in memory. However, streaming responses isn't as fast as sending `repeated` fields in a single message, so as a rule streaming shouldn't be used for small datasets.
+
+### Differences to WCF
+
+A WCF duplex service uses a client callback interface that can have multiple methods. A gRPC server streaming service can only send messages over a single stream. If you need multiple methods, use a message type with either [an Any field or a oneof field](protobuf-any-oneof.md) to send different messages, and write code in the client to handle them.
+
+In WCF, the one-way method called from the client returns immediately, but the [ServiceContract](https://docs.microsoft.com/dotnet/api/system.servicemodel.servicecontractattribute?view=netframework-4.8) class with the session is kept alive until the connection is closed. In gRPC, the `Task` returned by the implementation method shouldn't complete until the connection is closed.
```
Comment (mconnew) [inline docs/architecture/grpc-for-wcf-developers/rpc-types.md]: It doesn't return immediately as the server can still return an error. For example if you send a message for an action that the server doesn't recognize, a fault will be returned to the client and the client will throw an exception. So the server will verify that the message is for an operation it can dispatch and that the soap message is valid. For HTTP, the server replies with a 200 OK response with a 0 length content, for framing protocols such as NetTcp, it returns a single byte ACK. Only then does the client complete the Task/return from sync call.

File: docs/architecture/grpc-for-wcf-developers/rpc-types.md [hunk 25]
```diff
@@ -0,0 +1,268 @@
...
+
+## WCF one-way operations and gRPC client streaming
+
+WCF provides one-way operations (marked with `[OperationContract(IsOneWay = true)]`) that the client can call without waiting for a response. gRPC service methods always return a response, even if it's empty, and the client should always await that response. For "fire-and-forget" style messaging in gRPC, you can create a client streaming service.
```
Comment (mconnew) [inline docs/architecture/grpc-for-wcf-developers/rpc-types.md]: WCF waits for an ack (as mentioned in previous comment) which is sent just before dispatching. For WCF the difference between a void method and a one-way method is a void method returns an empty soap message once the operation is completed and a one-way method returns a transport specific ACK once the message has been verified as understood but before actual dispatch.

File: docs/architecture/grpc-for-wcf-developers/wcf-endpoints-grpc-methods.md [hunk 26]
```diff
@@ -0,0 +1,47 @@
...
+
+Chapter 3 showed that Protobuf message definitions are used to generate data classes. Service and method declarations are used to generate base classes that you inherit from to implement the service. You just declare the methods to be implemented in the `.proto` file, and the compiler generates a base class with virtual methods you must override.
+
+## OperationContract properties
+
+The [OperationContract](xref:System.ServiceModel.OperationContractAttribute) attribute has properties to control or refine how it works. gRPC methods don’t offer this type of control. The following table sets out those `OperationContract` properties and how the functionality they specify is (or isn’t) dealt with in gRPC:
+
+| `OperationContract` property | gRPC                                             |
+| ---------------------------- | ------------------------------------------------ |
+| <xref:System.ServiceModel.OperationContractAttribute.Action>             | SOAP-related, no meaning in gRPC. |
```
Comment (mconnew) [inline docs/architecture/grpc-for-wcf-developers/wcf-endpoints-grpc-methods.md]: Not exactly. It's a unique uri identifying the operation which can map to a SOAP header or even an HTTP header. How is an operation identified in gRPC? When a gRPC message arrives, how do you know which method the message is for?

File: docs/architecture/grpc-for-wcf-developers/wcf-endpoints-grpc-methods.md [hunk 27]
```diff
@@ -0,0 +1,47 @@
...
+| <xref:System.ServiceModel.OperationContractAttribute.AsyncPattern>       | All gRPC service methods return `Task` objects. |
+| <xref:System.ServiceModel.OperationContractAttribute.HasProtectionLevel> | No message encryption; network encryption handled at the transport layer (TLS over HTTP/2). |
```
Comment (mconnew) [inline docs/architecture/grpc-for-wcf-developers/wcf-endpoints-grpc-methods.md]: This isn't settable so isn't something you can control. It simply tells you if the ProtectionLevel value is the default or has been explicitly set.

File: docs/architecture/grpc-for-wcf-developers/wcf-endpoints-grpc-methods.md [hunk 28]
```diff
@@ -0,0 +1,47 @@
...
+| <xref:System.ServiceModel.OperationContractAttribute.IsInitiating>       | See [gRPC streaming](rpc-types.md#grpc-streaming). |
```
Comment (mconnew) [inline docs/architecture/grpc-for-wcf-developers/wcf-endpoints-grpc-methods.md]: I didn't see this addressed in the rpc-types.md file. This property basically basically is used to say a method can't be the first method called. The default is true so by default you can make any method call as your first call, but if you mark it as false, you must call some other method with this set as true first.

File: docs/architecture/grpc-for-wcf-developers/wcf-endpoints-grpc-methods.md [hunk 29]
```diff
@@ -0,0 +1,47 @@
...
+| <xref:System.ServiceModel.OperationContractAttribute.IsOneWay>           | One-way gRPC methods return `Empty` results or use client streaming. |
+| <xref:System.ServiceModel.OperationContractAttribute.IsTerminating>      | See [gRPC streaming](rpc-types.md#grpc-streaming). |
```
Comment (mconnew) [inline docs/architecture/grpc-for-wcf-developers/wcf-endpoints-grpc-methods.md]: I didn't see any reference to this in rpc-types.md. When an operation has this property set, it causes the server to initiate closing the session/connection (depending on transport) after it has sent the response to this operation. Can also be set on a callback contract and when the server calls that callback operation, the client will close the channel after the call completes.

File: docs/architecture/grpc-for-wcf-developers/ws-protocols.md [hunk 30]
```diff
@@ -0,0 +1,31 @@
...
+ms.date: 09/02/2019
+---
+
+# WS-\* protocols
+
+One of the real benefits of working with Windows Communication Foundation (WCF) was that it supported many of the existing _WS-\*_ standard protocols. This section will briefly cover how gRPC manages the same WS-\* protocols and discuss what options are available when there's no alternative.
+
+## Metadata Exchange - WS-Policy, WS-Discovery, and so on
+
+SOAP services expose Web Services Description Language (WSDL) schema documents with information such as data formats, operations, or communication options. This schema could be used to generate the client code.
```
Comment (mconnew) [inline docs/architecture/grpc-for-wcf-developers/ws-protocols.md]: WS-Discovery is used for more than a dynamic client. For example, although it doesn't use WS-Discovery, when you first run outlook, you provide your email address and it uses the domain part to do autodiscovery (using dns) to find your exchange server. So you don't need to use server reflection, you could have a client created from a known .proto but need to discover where your server which implements that .proto can be found. DNS based auto-discovery might be an alternative (which is what outlook/exchange actually uses).

File: docs/architecture/grpc-for-wcf-developers/ws-protocols.md [hunk 31]
```diff
@@ -0,0 +1,31 @@
...
+
+gRPC works best when servers and clients are generated from the same `.proto` files, but a server reflection optional extension does provide a way to expose dynamic information from a running server. For more information, see the [Grpc.Reflection](https://nuget.org/packages/Grpc.Reflection) NuGet package and the [gRPC C# Server Reflection](https://github.com/grpc/grpc/blob/master/doc/csharp/server_reflection.md) article.
+
+## Security – WS-Security, WS-Federation, XML Encryption, and so on
+
+Security, authentication, and authorization are covered in much more detail in [Chapter 6](authentication.md). But it's worth noting here that, unlike WCF, gRPC doesn't support WS-Security, WS Federation, or XML Encryption. Even so, gRPC offers greater security than WCF because all gRPC communications are automatically encrypted because of HTTP/2 TLS, and can use X509 certificates for mutual client/server authentication.
```
Comment (mconnew) [inline docs/architecture/grpc-for-wcf-developers/ws-protocols.md]: What's your basis for saying gRPC offers greater security than WCF? The reason why I don't like that sentence is it implies there are problems with WCF security, especially when Microsoft is the one publishing this book. Never mind the fact that there's a strong argument to be made that WCF provides greater security than gRPC. I could list a dozen security features that WCF provides which gRPC doesn't. The word "greater" is ambiguous.

File: docs/architecture/grpc-for-wcf-developers/ws-protocols.md [hunk 32]
```diff
@@ -0,0 +1,31 @@
...
+
+## WS-ReliableMessaging
+
+gRPC transport provides similar guarantees to WCF Reliable Sessions without any need for additional configuration or code.
```
Comment (mconnew) [inline docs/architecture/grpc-for-wcf-developers/ws-protocols.md]: It does not provide similar guarantees. I just spoke with the asp.net core team to verify and gRPC doesn't do any of the things that WS-ReliableMessaging provides.

File: docs/architecture/grpc-for-wcf-developers/ws-protocols.md [hunk 33]
```diff
@@ -0,0 +1,31 @@
...
+
+## WS-Transaction, WS-Coordination
+
+WCF’s implementation of distributed transactions uses Windows’ Microsoft Distributed Transaction Coordinator or MSDTC. It works with resource managers that specifically support it, like SQL Server, MSMQ, or Windows file systems. In the modern, open, cross-platform microservices world, there's no equivalent. For a discussion of transactions, see [Appendix A](appendix.md).
```
Comment (mconnew) [inline docs/architecture/grpc-for-wcf-developers/ws-protocols.md]: WS-Transaction is an open standard and WCF at least works with at least one Java based SOAP stack. It is open and cross-platform. It's not modern as nobody has reinvented that wheel yet.

File: docs/architecture/grpc-for-wcf-developers/ws-protocols.md (hunk 31 above)
Comment (markrendle) [reply docs/architecture/grpc-for-wcf-developers/ws-protocols.md]: Yeah, sorry about that, wasn't supposed to say "greater".

**Review by JamesNK — APPROVED** [review]

File: docs/architecture/grpc-for-wcf-developers/error-handling.md [hunk 34]
```diff
@@ -7,5 +7,65 @@ ms.date: 09/02/2019
...
+            { "User", user.Identity.Name }
+        };
+        throw new RpcException(new Status(StatusCode.PermissionDenied, "Permission denied"), metadata);
+    }
+}
+```
+
+## Catching errors in gRPC clients
+
+Just like WCF clients can catch <xref:System.ServiceModel.FaultException%601> errors, a gRPC client can catch an `RpcException` to handle errors. Since `RpcException` isn't a generic type, you can't catch different error types in different blocks, but you can use C#'s *exception filters* feature to declare separate `catch` blocks for different status codes, as shown in the following example:
```
Comment (mconnew) [inline docs/architecture/grpc-for-wcf-developers/error-handling.md]: Does the %601 form part of the xref? Just want to make sure this isn't accidental junk.

Comment (mairaw) [reply docs/architecture/grpc-for-wcf-developers/error-handling.md]: Yes. We need to encode the backtick in the xref links.

File: docs/architecture/grpc-for-wcf-developers/rpc-types.md [hunk 35]
```diff
@@ -0,0 +1,268 @@
...
+        {
+            Console.WriteLine(_stream.ResponseStream.Current.Text);
+        }
+    }
+
+    public async ValueTask DisposeAsync()
+    {
+        await _stream.RequestStream.CompleteAsync();
+        await _readTask;
+        _stream.Dispose();
```
Comment (mconnew) [inline docs/architecture/grpc-for-wcf-developers/rpc-types.md]: Should this be `await _stream.DisposeAsync();`? Not a blocker.

File: docs/architecture/grpc-for-wcf-developers/rpc-types.md [hunk 36]
```diff
@@ -0,0 +1,268 @@
...
+    }
+
+    public async Task SendAsync(string message)
+    {
+        await _stream.RequestStream.WriteAsync(new MessageRequest {Text = message});
+    }
+
+    private async Task ReadAsync(CancellationToken token)
+    {
+        while (await _stream.ResponseStream.MoveNext(token))
```
Comment (JamesNK) [inline docs/architecture/grpc-for-wcf-developers/rpc-types.md]: Use await foreach

File: docs/architecture/grpc-for-wcf-developers/rpc-types.md (hunk 35 above)
Comment (JamesNK) [reply docs/architecture/grpc-for-wcf-developers/rpc-types.md]: It's the call. It doesn't has an async dispose.

nit: In gRPC code samples I like to suffix call instances with call, e.g. chatCall. It makes it obvious what it is you're disposing.

File: docs/architecture/grpc-for-wcf-developers/ws-protocols.md [hunk 37]
```diff
@@ -0,0 +1,33 @@
...
+
+gRPC works best when servers and clients are generated from the same `.proto` files, but a server reflection optional extension does provide a way to expose dynamic information from a running server. For more information, see the [Grpc.Reflection](https://nuget.org/packages/Grpc.Reflection) NuGet package and the [gRPC C# Server Reflection](https://github.com/grpc/grpc/blob/master/doc/csharp/server_reflection.md) article.
+
+The WS-Discovery protocol is used to locate services on a local network. gRPC services are generally located using DNS or a service registry such as Consul or ZooKeeper.
+
+## Security – WS-Security, WS-Federation, XML Encryption, and so on
+
+Security, authentication, and authorization are covered in much more detail in [Chapter 6](authentication.md). But it's worth noting here that, unlike WCF, gRPC doesn't support WS-Security, WS Federation, or XML Encryption. Even so, gRPC provides excellent security; all gRPC network traffic is automatically encrypted when using HTTP/2 over TLS, and X509 certificates may be used for mutual client/server authentication.
```
Comment (mconnew) [inline docs/architecture/grpc-for-wcf-developers/ws-protocols.md]: Typo: WS-Federation.  
Also I thought gRPC can use WS-Federation, but the authentication token is used in a different way than WCF uses it. Maybe say "gRPC doesn't support WS-Security or WS-Federation with WS-Security" but then it starts getting a bit wordy and confusing. I'm not quite sure what the best way is to explain this. The key thing I think is if a customer today is using WS-Federation with WCF, then they can use WS-Federation with gRPC, it's just plumbed and configured differently.
Not a blocker though as this is getting into nuance.

File: docs/architecture/grpc-for-wcf-developers/ws-protocols.md [hunk 38]
```diff
@@ -0,0 +1,33 @@
...
+
+## WS-ReliableMessaging
+
+gRPC does not provide an equivalent to WS-ReliableMessaging. Retry semantics should be handled in code, possibly with a library like [Polly](https://github.com/App-vNext/Polly). When running in Kubernetes or similar orchestration environments, [service meshes](service-mesh.md) can also help to provide reliable messaging between services.
```
Comment (mconnew) [inline docs/architecture/grpc-for-wcf-developers/ws-protocols.md]:
```suggestion
gRPC does not provide an equivalent to WS-ReliableMessaging. Retry semantics should be handled in code, possibly with a library like [Polly](https://github.com/App-vNext/Polly). When running in Kubernetes or similar orchestration environments, [service meshes](service-mesh.md) can also help to provide reliability between services.
```
Use of the phrase "reliable messaging" suggests it's the same thing as that's tied to the WS-* protocol. The key thing is it helps with reliability.
Not a blocker for preview as again a nuance.

**Review by mconnew — APPROVED** [review]
Not requiring changes from my comments, nothing is a blocker. More feedback for the next refinement round.
